### DISCUSSION TITLE
Empty Address Dispensers

### DISCUSSION THREAD

**Discussion by B0BSmiths** [CounterpartyXCP/CIPs#138] [CIPs]
Counterparty Protocol is being changed to prevent opening dispensers on empty addresses, it is considered to be bug allowing anyone to impersonate someone else 

Empty addresses are used by many people for legitimate reasons, for example it allows one to open a dispenser on a burn address, a hardware wallet, or a p2sh address (multisig/OP_HODL etc etc), and the dispenser fixes are going to prevent this with a protocol breaking change. 

If ownership of an address can be proved then i would like to request it be allowed to be used as a dispenser.  One way to achieve this is to allow users to publish signed public keys or signed redeemScripts to prove ownership of an address prior to opening a dispenser on it.  This could be achieved by broadcasting the signed message from the address from which the dispenser is being opened from (source address)

Burn addresses should be permitted as they do not belong to anyone and proving a negative is very hard.

**Comment by adamkrellenstein** [comment]
So I think there's a misunderstanding: the issue under discussion (https://github.com/CounterpartyXCP/counterparty-core/issues/1785) would not prevent the creation of dispensers on empty address—it would simply prevent the creation of dispensers on addresses that you don't own, as you suggest. If you can prove you own the address, you can open a dispenser on it. No transaction history needed.

Unfortunately, I don't see how it's possible to allow for dispensers on burn addresses, at least as they exist today. If people want to allow burn-dispensing, we could add support for that independently.

**Reply by B0BSmiths** [reply]
Thank you that is good to know.  

Currently there is no mechanism to prove to Counterparty one owns a particular address, so the question arises how to prove address ownership?  Would, as I suggested, a signed public key or a signed redeemScript be sufficient to prove ownership? and would the idea of broadcasting this proof be sufficient for Counterparty to then access and verify the proof?

**Reply by adamkrellenstein** [reply]
Right now Counterparty transactions use the Bitcoin signature on the source address to prove ownership of that address.

**Reply by B0BSmiths** [reply]
My apologies I mean to say Currently there is no mechanism to prove to Counterparty one owns a particular empty as in unused address

**Reply by adamkrellenstein** [reply]
Yep, I'm not sure what method is best. I think it probably makes sense to start with using `source` just like for other transactions. It's easier to use `source` than manually produce a signature of another kind. Note, this wouldn't preclude use of hardware wallets etc.

**Reply by B0BSmiths** [reply]
As I understand the create_dispenser api method 'source' would need to be the address that currently possess the assets that are to be dispensed and "open_address" is perhaps where we could pass a signed pubkey or a signed redeemScript to the API so that it can verify it and build a txhex.  

We will need to include the proof in a Counterparty message within the tx that is then broadcast, so that all Counterparty nodes can also verify it after it has been mined, so that all nodes open the dispenser on the empty address on which ownership is proven.  

OP_RETURN is limited in data carrying capacity and P2MS is very expensive as it results in transactions with outputs in which 33% is the users dust return pubkey repeated many times.  P2SH is a two transaction affair so perhaps a new encoding mechanism based on the ordinal envelope would be a good way to achieve it in a single tx and would allow users to take advantage of the segwit discount
